### PR TITLE
Expose qos durability and reliability to ros2topic echo

### PR DESCRIPTION
**Pull request by emersonknapp**
Expose the same QoS settings to `ros2 topic echo` as were exposed to `ros2 topic pub` in https://github.com/ros2/ros2cli/pull/238

Fixes https://github.com/ros2/ros2cli/issues/281
Closes https://github.com/ros-security/aws-roadmap/issues/57

Blocked by #304

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
@emersonknapp what's the status of this PR. Is this something that will be resumed or should it be closed and left as a future enhancement?

**Comment by zmichaels11** [comment]
@nuclearsandwich Emerson is on vacation until Aug 20th. 
Can this wait until then?

**Comment by emersonknapp** [comment]
@nuclearsandwich thanks for following up - this fell off my radar. I'm going to fix this PR up and open it for review early next week

**Comment by emersonknapp** [comment]
@mjcarroll I have updated this PR with some sanity check pytests. Please take a look! :)

I'm a little worried about compatibility with this big change https://github.com/ros2/ros2cli/pull/304 - it's possible I've duplicated some small portion of what's going on in there. Though I doubt it will be in conflict.

File: ros2topic/ros2topic/verb/echo.py [hunk 1]
```diff
@@ -112,10 +131,9 @@ def subscriber(node, topic_name, message_type, callback):
     msg_module = import_message_type(topic_name, message_type)
 
     node.create_subscription(
-        msg_module, topic_name, callback, qos_profile_sensor_data)
+        msg_module, topic_name, callback, qos_profile)
 
-    while rclpy.ok():
-        rclpy.spin_once(node)
+    rclpy.spin(node)
```
Comment (piraka9011) [inline ros2topic/ros2topic/verb/echo.py]: Any reason for removing the while loop?

Comment (rotu) [reply ros2topic/ros2topic/verb/echo.py]: There's a good reason. creating an executor, which rclpy does, is insanely expensive. We're talking like 30 ms of cpu work for some reason.

Comment (emersonknapp) [reply ros2topic/ros2topic/verb/echo.py]: Part of what I was trying to do here was put some consistency between pub and echo. `spin` exists for this exact reason anyways and it's what `pub.py` does

Comment (rotu) [reply ros2topic/ros2topic/verb/echo.py]: https://discourse.ros.org/t/singlethreadedexecutor-creates-a-high-cpu-overhead-in-ros-2/10077

Comment (piraka9011) [reply ros2topic/ros2topic/verb/echo.py]: Thanks!

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8394/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4308/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6834/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8293/)

**Comment by emersonknapp** [comment]
Ok, just as long as we get it in for the feature freeze! :) thanks

**Comment by mjcarroll** [comment]
Alright, no longer waiting on #304, so once Windows CI goes green, I'm happy to merge.

**Comment by emersonknapp** [comment]
It seems like the Windows build is failing to spawn Python subprocesses, period. https://ci.ros2.org/job/ci_windows/8354/testReport/junit/ros2topic.test/test_xmllint/test_xmllint/ e.g. xmllint is failing. Any insight on that?

**Comment by emersonknapp** [comment]
The issue may be related to this python bug that affected windows python https://bugs.python.org/issue37380 - perhaps we haven't gotten a fixed version on this build machine

**Comment by emersonknapp** [comment]
Running windows build with subprocess-based tests disabled [Build Status](https://ci.ros2.org/job/ci_windows/8363/)

**Comment by mjcarroll** [comment]
Great! I'm going to merge this, but can you add an issue to follow up to resolve the tests (which may involve changing python versions on the Windows CI boxes)

**Comment by dirk-thomas** [comment]
> Running windows build with subprocess-based tests disabled

@emersonknapp While this has been merged now I only want to state that this is not the idea of the API/feature freeze deadline. In the future please make sure that you get PRs early enough into shape that they pass across all target platforms. Otherwise the changes might simply not make it in time for an upcoming release and get bumped to the next ROS distro instead.

**Comment by emersonknapp** [comment]
@dirk-thomas when do we plan to update the python version on the windows build machines? This has nothing to do with my code, I only exposed an existing problem in the build infrastructure that anybody could have run into

**Comment by dirk-thomas** [comment]
Ok, lets hope this is being fixed by a new Python version. As far as I can see there is no 3.7.x release containing the fix yet. We considered updating to 3.8 but haven't done so last minute before the freeze date (see https://github.com/ros2/build_cop/issues/243#issuecomment-543340758). We will likely try it next week and see if there are any issues with it (and if it resolves this case).

**Comment by emersonknapp** [comment]
I believe 3.7.5 contains the fix but I'm not positive

**Comment by rotu** [comment]
It looks like the subprocess crash is just a tack-on effect and the root cause lies in the rclpy library. The first error is from an "Unsupported signal: 2" being raised from `rclpy.shutdown`. This is calling `subprocess.send_signal` which doesn't understand SIGINT.

https://github.com/python/cpython/blob/master/Lib/subprocess.py#L1516

```
20:26:52 _____________ test_pub_basic[/clitest/topic/pub_basic-False-True] _____________
20:26:52 test\test_echo_pub.py:103: in test_pub_basic
20:26:52     executor.spin_until_future_complete(future, timeout_sec=3)
20:26:52 ..\..\..\..\install\Lib\site-packages\rclpy\executors.py:297: in spin_until_future_complete
20:26:52     self.spin_once(timeout_sec=timeout_left)
20:26:52 ..\..\..\..\install\Lib\site-packages\rclpy\executors.py:684: in spin_once
20:26:52     raise handler.exception()
20:26:52 ..\..\..\..\install\Lib\site-packages\rclpy\task.py:210: in __call__
20:26:52     self._handler.send(None)
20:26:52 ..\..\..\..\install\Lib\site-packages\rclpy\executors.py:404: in handler
20:26:52     await call_coroutine(entity, arg)
20:26:52 ..\..\..\..\install\Lib\site-packages\rclpy\executors.py:330: in _execute_subscription
20:26:52     await await_or_execute(sub.callback, msg)
20:26:52 ..\..\..\..\install\Lib\site-packages\rclpy\executors.py:118: in await_or_execute
20:26:52     return callback(*args)
20:26:52 test\test_echo_pub.py:97: in message_callback
20:26:52     shutdown()
20:26:52 test\test_echo_pub.py:90: in shutdown
20:26:52     process.send_signal(signal.SIGINT)
20:26:52 c:\python37\lib\subprocess.py:1306: in send_signal
20:26:52     raise ValueError("Unsupported signal: {}".format(sig))
20:26:52 E   ValueError: Unsupported signal: 2
20:26:52 _________ test_pub_basic[/clitest/topic/pub_compatible_qos-True-True] _________
20:26:52 test\test_echo_pub.py:87: in test_pub_basic
20:26:52     stdin=subprocess.DEVNULL)
20:26:52 c:\python37\lib\subprocess.py:664: in __init__
20:26:52     _cleanup()
20:26:52 c:\python37\lib\subprocess.py:228: in _cleanup
20:26:52     res = inst._internal_poll(_deadstate=sys.maxsize)
20:26:52 c:\python37\lib\subprocess.py:1216: in _internal_poll
20:26:52     if _WaitForSingleObject(self._handle, 0) == _WAIT_OBJECT_0:
20:26:52 E   OSError: [WinError 6] The handle is invalid
20:26:52 ____________ test_echo_basic[/clitest/topic/echo_basic-False-True] 
```

https://ci.ros2.org/job/ci_windows/8354/console#console-section-11

**Comment by rotu** [comment]
Oops I interpreted that slightly wrong. While rclpy is in the callstack, the send_signal(SIGINT) call is in test_echo_pub.py:
https://github.com/ros2/ros2cli/blob/bb555903c8260b39388df3c5f7bece3bf6d89c5b/ros2topic/test/test_echo_pub.py#L92

**Comment by emersonknapp** [comment]
Yeah on further look it seems that Windows doesn't like the SIGINT - It seems like that may be what caused the subsequent tests to fail to spin up subprocesses at all by triggering the situation for the Python subprocess bug?

It seems like using the framework/pattern set being introduced here https://github.com/ros2/ros2cli/pull/304 would be more valuable than fixing this test as-is (for consistency in this package), so I propose I refactor this test to build on that, once it is merged.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/inconsistency-in-flake8-between-ros2-source-builds-ci-ros2-org-and-build-ros2-org-for-prs/11079/1

**Pull request merged**